Source/WebGPU/WebGPU/CommandEncoder.swift

```
@@ -1183,8 +1183,11 @@ extension WebGPU.CommandEncoder {
         if attachmentsToClear.count != 0 || depthStencilAttachmentToClear != nil {
             let attachment = wgpuGetRenderPassDescriptorDepthStencilAttachment(descriptorSpan)?[0]
             if attachment != nil && depthStencilAttachmentToClear != nil {
```

this should use `if let`, and then no force unwrapping needs to happen, and no potential crashes can happen

---

We can write the code like so:
```
            if depthStencilAttachmentToClear != nil {
                if let attachment {
                    var textureView: WebGPU.TextureOrTextureView
                    if let view = attachment.view {
                        textureView = WebGPU.TextureOrTextureView(WebGPU.fromAPI(view))
                    } else if let texture = attachment.texture {
                        textureView = WebGPU.TextureOrTextureView(WebGPU.fromAPI(texture))
                    }
                }
                textureView.setPreviouslyCleared()
            }
```

but I'm not sure its more readable. 

Safe crashing is very much preferred over incorrect logic which doesn't crash.

---

what about just this (inside the depthStencilAttachmentToClear check thing)

```
let textureView = if let view = attachment?.view { 
   WebGPU.TextureOrTextureView(WebGPU.fromAPI(view))
else if let texture = attachment?.texture {
   WebGPU.TextureOrTextureView(WebGPU.fromAPI(texture))
}
 ```
 
 I _strongly_ think that force unwraps should _always_ be avoided, since they essentially suppress the compiler and turn compile time issues into runtime crashes

---

in fact, you could even do something like this!

```
extension WebGPU.TextureOrTextureView {

init?(attachment: Attachment?) {
 if let view = attachment?.view { 
   self.init(WebGPU.fromAPI(view))
else if let texture = attachment?.texture {
   self.init(WebGPU.fromAPI(texture))
} else {
  return nil
}

}
```

and then you can just do `let textureView = WebGPU.TextureOrTextureView(attachment)`

---

endless possibilities of making it easily readable and comprehensible while leveraging Swift’s strengths and not going against the compiler :)

---

> I strongly think that force unwraps should always be avoided, since they essentially suppress the compiler and turn compile time issues into runtime crashes

+1, force unwrapping is an anti-pattern.

---

Can you please explain how it's different from dereferencing a RefPtr (specifically calling `RefPtr::operator*()`)?

I agree we should avoid it, but some of the patterns suggest move runtime crashes into silent failures which are much worse.

---

> Can you please explain how it's different from dereferencing a RefPtr (specifically calling RefPtr::operator*())?

C++ as a language doesn’t have compiler support for optional values, so runtime checking for null-ness and then dereferencing is the best we can do. But Swift _does_ have compile time support for these, so it can and should be leveraged.

---

There is a very large risk to introducing silent failures with improper `if let` and `guard let` usage which in GPU process code can be worse than crashing. 

I do like the pattern where it makes sense. The argument that it should be used exclusively is not strong and I think subject to discussion on webkit-dev before any formal policies are in place.

---

> There is a very large risk to introducing silent failures with improper if let and guard let usage

Hm, do you have any concrete examples of how these might introduce silent failures compared to using a force unwrap? In general, going against the compiler and suppressing a form of type checking should usually be avoided if possible.

---

> this should use if let, and then no force unwrapping needs to happen, and no potential crashes can happen

> +1, force unwrapping is an anti-pattern.

Where is the force-unwrapping? I see an optional passed to `WebGPU.TextureOrTextureView`, which accepts an optional and then uses `if let` to unwrap it.

---

@geoffreygaren this was prior to the latest version of this PR, so this comment thread is no longer directly applicable

---

> > There is a very large risk to introducing silent failures with improper if let and guard let usage
> 
> Hm, do you have any concrete examples of how these might introduce silent failures compared to using a force unwrap? In general, going against the compiler and suppressing a form of type checking should usually be avoided if possible.

Yes a great example is this bug :) It would have never been detected had we said:
```
if let view = attachment.view {
   view.setPreviouslyCleared()
}
```
it was only detected because we force unwrapped it like:
```
view!.setPreviouslyCleared()
```

I don't really understand the going against the compiler argument when force-unwrapping is a language feature supported b the compiler :) And in this case it led to the discovery of an issue we would have otherwise likely not discovered until a 3rd party reported it, the CTS expanded coverage in this area, or our own tests expanded coverage here

---

I tend to agree with @mwyrzykowski here that `view!` is clearer than `if let view { } else { fatalError() }`. @rr-codes @pxlcoder can you explain why you find the second form clearer? Or were you referring to a general principle that might not apply in this specific case?

---

We can continue to discuss, but I think this is outside the scope of this PR.